metaflow/plugins/argo/event_decorators.py

```diff
@@ -0,0 +1,247 @@
+
+
+class TriggerOnDecorator(FlowDecorator):
+
+    name = "trigger_on"
+    defaults = {
+        "flow": None,
+        "status": "succeeded",
+        "event": None,
+        "data": None,
+        "mappings": None,
+    }
+    options = {
+        "flow": dict(
+            is_flag=False,
+            show_default=True,
+            help="Trigger the current flow when this flow completes.",
+        ),
+        "event": dict(
+            is_flag=False,
+            show_default=True,
+            help="Trigger the flow when this event is received",
+        ),
+        "status": dict(
+            is_flag=False,
+            show_default=True,
+            help="Status of triggering flow. Valid values are 'succeeded', 'failed'",
+        ),
+        "mappings": dict(
+            is_flag=False,
+            show_default=True,
+            help="Mapping of event fields to flow parameters",
+        ),
+    }
+
+    def flow_init(
+        self, flow, graph, environment, flow_datastore, metadata, logger, echo, options
+    ):
+        if are_events_configured():
+            self._option_values = options
+            flow_name = self.attributes.get("flow")
+            event_name = self.attributes.get("event")
+            if flow_name in [None, ""] and event_name in [None, ""]:
+                raise MetaflowException(
+                    msg=("@%s needs a flow or event name." % self.name)
+                )
+            trigger_on = {
+                "flow": flow_name,
+                "status": None,
+                "event": event_name,
+                "mappings": None,
+            }
+            if flow_name is not None:
+                status = self.attributes.get("status").lower()
+                if status not in valid_statuses():
+                    raise MetaflowException(
+                        msg=(
+                            "@%s requires status to be one of: %s."
+                            % (self.name, ", ".join(valid_statuses()))
+                        )
+                    )
+                trigger_on["status"] = status
+            elif event_name is not None:
+                sanitized = sanitize_user_event(event_name)
+                if sanitized == "":
+                    raise BadEventNameException(BAD_EVENT_ERR_MSG)
+                trigger_on["event"] = sanitized
+            mappings = self.attributes.get("mappings")
+            trigger_on["mappings"] = validate_mappings(mappings)
+            current._update_env({"trigger_on": trigger_on})
+
+        else:
+            # Defer raising an error in case user has specified --ignore-triggers
+            current._update_env(
+                {
+                    "event_decorator_error": {
+                        "message": ERR_MSG,
+                        "headline": ("@%s requires eventing support" % self.name),
+                    }
+                }
+            )
+
+
+class EmitEventDecorator(StepDecorator):
+
+    name = "emit_event"
```

Comment:
rather than a decorator, we can make this a function. That way the user retains more control of the payload going into the event (it can be a function of user code within a step).

```
@step
def foo(self):
    x = get_some_output()
    event.emit(..., x, ...)
```